kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaAuthnHandler.java

```diff
@@ -0,0 +1,403 @@
+                data.correlationId(),
+                new ResponseHeaderData().setCorrelationId(data.correlationId()),
+                new SaslHandshakeResponseData()
+                        .setMechanisms(enabledMechanisms)
+                        .setErrorCode(error.code())));
+
+    }
+
+    private void onSaslAuthenticateRequest(ChannelHandlerContext ctx,
+                                           DecodedRequestFrame<SaslAuthenticateRequestData> data) {
+        byte[] bytes = new byte[0];
+        Errors error;
+        String errorMessage;
+
+        try {
+            bytes = doEvaluateResponse(ctx, data.body().authBytes());
+            error = Errors.NONE;
+            errorMessage = null;
+        }
+        catch (SaslAuthenticationException e) {
+            error = Errors.SASL_AUTHENTICATION_FAILED;
+            errorMessage = e.getMessage();
+        }
+        catch (SaslException e) {
+            error = Errors.SASL_AUTHENTICATION_FAILED;
+            errorMessage = "An error occurred";
+        }
+
+        ctx.writeAndFlush(
+                new DecodedResponseFrame<>(
+                        data.apiVersion(),
+                        data.correlationId(),
+                        new ResponseHeaderData().setCorrelationId(data.correlationId()),
+                        new SaslAuthenticateResponseData()
+                                .setErrorCode(error.code())
+                                .setErrorMessage(errorMessage)
+                                .setAuthBytes(bytes)
+                // TODO add support for session lifetime
+                ));
+    }
+
+    private boolean supportsSaslGssApi() {
+        return false;
+    }
+
+    private byte[] doEvaluateResponse(ChannelHandlerContext ctx,
+                                      byte[] authBytes)
+            throws SaslException {
+        final byte[] bytes;
+        try {
+            bytes = saslServer.evaluateResponse(authBytes);
+        }
+        catch (SaslAuthenticationException e) {
+            validateTransition(State.FAILED);
+            saslServer.dispose();
+            throw e;
+        }
+        catch (Exception e) {
+            validateTransition(State.FAILED);
+            saslServer.dispose();
+            throw new SaslAuthenticationException(e.getMessage());
+        }
+
+        if (saslServer.isComplete()) {
+            validateTransition(State.AUTHN_SUCCESS);
+            String authorizationId = saslServer.getAuthorizationID();
+            String[] properties = SaslMechanism.fromMechanismName(saslServer.getMechanismName()).negotiableProperties();
+            Map<String, Object> negotiatedProperty = new HashMap<>((int) (properties.length * 4.0 / 3));
+            for (String property : properties) {
+                Object value = saslServer.getNegotiatedProperty(property);
+                if (value != null) {
+                    negotiatedProperty.put(property, value);
+                }
+            }
```

Comment:
Great idea: It allows me to get rid of the temp HashMap entirely, using a `Map.of` instead. The `Map.copyOf` in the `AuthenticationEvent` then ends up returnings the same map, since it know's there's no need to actually copy an immutable map. Thanks!